### PR TITLE
Time range picker

### PR DESCRIPTION
**Pull request by meluru**
improve date picker as well

### CODE REVIEW

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+function _humanizeDuration(duration) {
+    const hours = duration.hours();
+    const minutes = duration.minutes();
+    if (hours === 1 && minutes === 0) {
+        return `${hours} hour`;
+    } else if (hours !== 0) {
+        const fraction = minutes / 60;
+
+        return `${hours + fraction} hours`;
```
Comment (mic4ael) [inline indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: You can limit it to only one line ``return `${hours + (minutes / 60)} hours`;``

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+    generateEndTimeOptions = (start) => {
+        const options = [];
+        const end = moment().endOf('day');
+        const next = moment(start).add(30, 'm');
+        let serializedNext, duration, text;
+        while (next < end) {
+            duration = _humanizeDuration(moment.duration(next.diff(start)));
+            serializedNext = _serializeTime(moment(next));
+            text = <span>{serializedNext} <span styleName="duration">({duration })</span></span>;
```
Comment (mic4ael) [inline indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: extra space in `{duration }`

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    const hours = duration.hours();
+    const minutes = duration.minutes();
+    if (hours === 1 && minutes === 0) {
+        return `${hours} hour`;
+    } else if (hours !== 0) {
+        const fraction = minutes / 60;
+
+        return `${hours + fraction} hours`;
+    } else {
+        return `${minutes} minutes`;
```
Comment (mic4ael) [inline indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: i guess it should be a translated string (like the others above)

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+            duration = _humanizeDuration(moment.duration(next.diff(start)));
+            serializedNext = _serializeTime(moment(next));
+            text = <span>{serializedNext} <span styleName="duration">({duration })</span></span>;
+            options.push({key: serializedNext, value: serializedNext, text});
+            next.add(30, 'm');
+        }
+        return options;
+    };
+
+    addOptionIfNotExist = (options, value) => {
```
Comment (mic4ael) [inline indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: maybe a better name would be `addOptionIfMissing`?

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+                duration,
+                startOptions,
+                endOptions,
+            });
+            const {onChange} = this.props;
+            onChange(start, end);
+        }
+    };
+
+    searchTime = () => {
```
Comment (mic4ael) [inline indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: what is the point of this function when passed to `Dropdown`?

Comment (meluru) [reply indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: It overrides the default behaviour of search, so when you start typing you get no options

Comment (mic4ael) [reply indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: yeah, yeah, but isn't it enough to not include the `search` attribute in the first place? or at least do it as `search={true}`. I don't understand why you pass the `search` prop even if you don't need the searching functionality

Comment (meluru) [reply indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: So basically including search property let's you type in the field and it's required by addition properties. Keeping the default behavior of search would try to match current options with what you type, so if you started typing e.g. '1' you'd get all the time options that include 1 which doesn't make much sense in a time picker. That's why I need this method to make sure that whatever you type, you get no options as a result of search.

Comment (mic4ael) [reply indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: Ahh, that is what I didn;t understand - `search` being required by some other properties. However you can make it an inline arrow function `search={() => []}`

Comment (meluru) [reply indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: that's a good idea :)

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx (hunk 1 above)
Comment (meluru) [reply indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: Are you sure I need parentheses here?

Comment (mic4ael) [reply indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: actually you don't thanks to the operator precedence but I think it is usually clearer that way (at least to me)

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    } else if (hours !== 0) {
+        return Translate.string('{time} hours', {time: hours + (minutes / 60)});
+    } else {
+        return Translate.string('{time} minutes', {time: minutes});
+    }
+}
+
+export default class TimeRangePicker extends React.Component {
+    static propTypes = {
+        startTime: propTypes.object,
```
Comment (mvidalgarcia) [inline indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: The way to go pointed out by the [`prop-types`](https://www.npmjs.com/package/prop-types) lib is importing it capitalizing the first letter i.e `PropTypes`. I know we have it imported as `propTypes` in other places of our code base but we should choose one way and be consistent all over.

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+import React from 'react';
+import propTypes from 'prop-types';
+import {Dropdown} from 'semantic-ui-react';
+import {Translate} from 'indico/react/i18n';
+import {toMoment} from '../util';
+
+import './TimeRangePicker.module.scss';
+
+
+const _serializeTime = time => (time ? time.format('HH:mm') : null);
```
Comment (mvidalgarcia) [inline indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: I've noticed we have the exact same function also in Landing.jsx and TimeForm.jsx, what about moving it to util and reusing it?

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx (hunk 1 above)
Comment (ThiefMaster) [reply indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: yeah, in such a case I prefer parentheses or not putting space around `/` if eslint allows that (PEP8 does, not sure about eslint)

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+import moment from 'moment';
+import React from 'react';
+import propTypes from 'prop-types';
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: Let's use the correct style in new files as @mvidalgarcia mentioned below. Existing files... either wait for the eslint-based fixer or update when changing something related.

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+        });
+        if (!found) {
+            options.unshift({key: value, value, text: value});
+            return true;
+        }
+        return false;
+    };
+
+    updateStartTime = (startTime, endTime, duration) => {
+        const start = moment(startTime, ['HH:mm', 'HHmm', 'Hmm']);
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: should we except `HH` as well? so writing just `13` would turn into `13:00`

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        let end = toMoment(endTime, 'HH:mm');
+        if (start.isValid()) {
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]:
```
if (!start.isValid()) {
    return;
}
```

File: indico/modules/rb_new/client/js/components/pages/Landing.jsx [hunk 11]
```diff
@@ -19,21 +19,22 @@ import moment from 'moment';
 import PropTypes from 'prop-types';
 import React from 'react';
 import {Link} from 'react-router-dom';
-import Calendar from 'rc-calendar';
+import RcCalendar from 'rc-calendar';
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/components/pages/Landing.jsx]: `RCCalendar`?

File: indico/modules/rb_new/client/js/components/pages/Landing.jsx [hunk 12]
```diff
@@ -204,11 +212,9 @@ export default class Landing extends React.Component {
...
                                 )}
                                 <Form.Group inline>
-                                    <TimePicker {...timePickerProps} value={startTime}
-                                                onChange={(value) => this.updateTimes(value, endTime)} />
-                                    -
-                                    <TimePicker {...timePickerProps} value={endTime}
-                                                onChange={(value) => this.updateTimes(startTime, value)} />
+                                    <TimeRangePicker startTime={startTime}
+                                                     endTime={endTime}
+                                                     onChange={(start, end) => this.updateTimes(start, end)} />
```
Comment (ThiefMaster) [inline indico/modules/rb_new/client/js/components/pages/Landing.jsx]: `onChange={this.updateTimes}`

File: indico/modules/rb_new/client/js/components/TimeRangePicker.jsx (hunk 9 above)
Comment (meluru) [reply indico/modules/rb_new/client/js/components/TimeRangePicker.jsx]: The first formatting also accepts HH

**Pull request closed without merging**